best-practices/pair-programming.md

```diff
@@ -0,0 +1,155 @@
+  hand and the solution which makes for better thought-through code.
+- knowledge sharing. Both in terms of programming knowledge in general
+  (mentoring), and within the application / system that is being developed
+  (avoiding silos), but also with regards to good habits and team practices.
+
+## How do we pair program?
+### Prerequisites
+For pair programming we use the following setup:
+- a keyboard for each developer
+- a trackpad or mouse for each developer
+- a shared screen, either the same physical one or with screen sharing
+
+### Pair programming styles
+There are different styles of pair programming, each with their own pros and
+cons.
+
+The styles described in this document are:
+- Ping Pong
+- Driver and Navigator
+- Tour Guide and Tourist
+- Strong style / Backseat Driver / Driving School
+- Unstructured
+
+#### Ping Pong
+**When to use: When the pair wants to ensure that their code is test-driven.**
+
+The Ping Pong style is recognized by its constant role switching where the pair
+switches in a red-green fashion.
+
+- Developer 1 writes a failing test.
+- Developer 2 writes just enough code to make the test pass.
+- Developer 2 writes a new failing test.
+- Developer 1 writes just enough code to make the test pass. 
+- This back-and-forth continues until there are no more tests for the unit they
+  are working on. 
+- They finish by refactoring if needed. 
+
+Then the cycle repeats with the next unit.
+
+The current Navigator works as sparring for the Driver to help them either write
+the test or solve the test.
+
+This style forces the pair to switch roles regularly which can help keep both
+developers focused for the session. It is especially useful for enforcing a
+test-driven approach.
+
+#### Driver and Navigator
+**When to use: When Ping Pong pairing is too taxing.**
+
+Driver and Navigator style can be seen as the most balanced pairing style, and
+this is also the style that the "Tour Guide and Tourist" and "Backseat Navigator
+/ Driver School / Strong Style" derives from.
+
+The pair gets assigned the role of either the Driver or the Navigator.
+
+The Driver has their fingers on the keyboard, working on the task at hand while
+discussing the solution with the Navigator.
+
+You might hear the Driver ask questions such as "What do you think is the best
+approach here?"
+
+The Navigator has an overview over the current task and its relation to the
+entire project. They work as a sparring partner to the Driver.
+
+You might hear the Navigator ask questions such as "Would it be useful to
+extract this method?"
+
+The Driver needs to be aware of not taking over the pairing session, coding
+solo, which would be reminiscent of the Tour guide and Tourist style.
+
+The Navigator needs to be aware of not micro-managing, which would be
+reminiscent of the Backseat Navigator style. 
+
+This awareness is especially important when one party has more experience.
+
```

Comment:
```suggestion

I can be beneficial to agree on a schedule for switching roles. For instance every 30 minutes.
```

Reply:
Good idea!

Reply:
We'll add a paragraph about it later today

Reply:
Agree. Maybe a good opportunity to nudge Pomodoro?